server/src/main/java/org/elasticsearch/transport/InboundDecoder.java

```diff
@@ -0,0 +1,241 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.transport;
+
+import org.elasticsearch.Version;
+import org.elasticsearch.common.bytes.BytesReference;
+import org.elasticsearch.common.bytes.ReleasableBytesReference;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.lease.Releasable;
+import org.elasticsearch.common.util.PageCacheRecycler;
+import org.elasticsearch.core.internal.io.IOUtils;
+
+import java.io.IOException;
+import java.util.function.Consumer;
+
+public class InboundDecoder implements Releasable {
+
+    static final Object PING = new Object();
+    static final Object END_CONTENT = new Object();
+
+    private final Version version;
+    private final PageCacheRecycler recycler;
+    private Exception decodingException;
+    private TransportDecompressor decompressor;
+    private int totalNetworkSize = -1;
+    private int bytesConsumed = 0;
+    private boolean isClosed = false;
+
+    public InboundDecoder(Version version, PageCacheRecycler recycler) {
+        this.version = version;
+        this.recycler = recycler;
+    }
+
+    public int decode(ReleasableBytesReference reference, Consumer<Object> fragmentConsumer) throws IOException {
+        ensureOpen();
+        try {
+            return internalDecode(reference, fragmentConsumer);
+        } catch (IOException e) {
```

Comment:
should we catch `Exception` here to make sure we're always cleaning up the decompressor for any kind of runtime exception as well?